Un-inline 'string' to fix throwing FS0670, and remove null-check

abelbraaksma opened the pull request:
Fixes #7958, #9153

In summary, this fixes the following things:

* When `string` was used on a variable inside a generic type, where that variable has that generic type, it failed with FS0670. This disallowed its use in many situations, like in `override this.ToString()` (see #7958)
* The original code had a redundant (although invisible) null-check (see #9153), it is now removed
* The original code used compiler-syntax that was unreachable, leading to the code always falling-through to use `anyToString "" value`. The dead code is removed.
* Added tests, specifically also to ensure that `InvariantCulture` is definitely used (I added tests that switch the culture to check if it influences the result)

After this change:

* The `string` function can be used anywhere where a normal generic function can be used (exception: ref types, see #9402)
* The original special-casing in the `string` function was never called or compiled, though even if it had been, it would've lead to the same result
* No effect on performance for as far as I can tell. Measurements are within error range. There's one extra `call`, but also one less null-check.

abelbraaksma commented:
> Expected: ["type OperatorTestsByte";
> "let testByteEqualsOperator(e1) (e2) = Operators.op_Equality<Microsoft.FSharp.Core.byte> (e1,e2) @ (4,63--4,72)";
> "let testByteNotEqualsOperator(e1) (e2) = Operators.op_Equality<Microsoft.FSharp.Core.bool>(Operators.op_Equality<Microsoft.FSharp.Core.byte> (e1,e2),False) @ (5,63--5,73)";

That's unexpected! What are these tests? Doesn't sound like they rely on `string` in any way.

abelbraaksma commented:
It also throws this:

Which in turn appears to be a bug in the runtime, or at least a resurfacing of a bug, possibly related are:

* https://stackoverflow.com/questions/5634417/caught-exception-is-null-itself?noredirect=1&lq=1
* https://github.com/dotnet/runtime/issues/5203
* https://github.com/dotnet/runtime/issues/6299
* https://github.com/dotnet/runtime/issues/6283

abelbraaksma commented:
Ok, that last error comes from the temp `exe` that is created, which, when run in isolation, shows the actual error:

> An unhandled exception of type 'System.IO.FileNotFoundException' occurred in Unknown Module.
> Could not load file or assembly 'System.Runtime, Version=4.2.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.

/cc @KevinRansom, help?

cartermp commented:
Those parts of the FCS tests are horribly annoying and not very simple to fix. The FCS tests also don't use the built FSharp.Core since FCS is currently designed to be compatible with dowlevel FSharp.Core binaries. I personally don't like that behavior, but it is what it is.

abelbraaksma commented:
>  The FCS tests also don't use the built FSharp.Core since FCS is currently designed to be compatible with dowlevel FSharp.Core binaries

Thanks, @cartermp, I just noticed that the actual tests that were failing here had zero dependency on FSharp.Core. Since they used `string`, which used to be inlined, that was not a problem. I've now changed these calls to `ToString()` instead, so that I don't have to change the whole config of these tests to use the current FSharp.Core.

The other ones, from this comment above (https://github.com/dotnet/fsharp/pull/9549#issuecomment-648525589), I've no idea yet, as I don't see them locally (or I'm doing something wrong). I'm now running all tests locally again, see if I missed something.

abelbraaksma commented:
@KevinRansom, in our discussion on this, you asked if we could improve, esp. to prevent excessive boxing. My take on this at the time was: no, we can't combine statically-resolvable type compiler syntax and at the same be non-inline (to prevent FS0670 from throwing).

But after some experimenting, and searching for uses in the source code of the `(# ..#)` inline IL syntax and static `when` clauses, it turns out that the compiler understands what to do with _normal_ generic types like `'T` and will inline when you give it an instruction to do so on a specific path, regardless of the function being marked `inline` or not. Pretty amazing and powerful stuff! This syntax may help in some other scenarios as well.

Since this idea makes this change larger than originally intended, I decided to go the "long" route and write out an RFC (see https://github.com/fsharp/fslang-design/pull/479) and the corresponding lang suggestion (see https://github.com/fsharp/fslang-suggestions/issues/890), which basically backs the findings of several bug reports. It's a minimal change in the surface area, but a change after all, hence the RFC.

So where does that leave us in this PR? Well, in short, we now have the best of both worlds:

* `string` is no longer inline, it won't throw FS0670 anymore (it still shows it in the editor, but it compiles fine now).
* IL generation for all basic struct types is now without copying, no boxing
* IL generation for _integral numeric types_ still requires boxing, because these can be used for `Enum` and need to be runtime-tested for `IFormattable`. It would be nice if I could distinguish statically between an enum and an int, but there's presently no way that works.
* I removed null-checks for other structs, as structs cannot be null
* Added no-op for `string` and fast-paths for `decimal`, `char` and a few others.

All in all, I'm pretty happy with the result in terms of generated IL. Maybe @teo-tsirpanis could also have a second-opinion look at this, as he reported the "bad IL" issue (https://github.com/dotnet/fsharp/issues/9153).

Note: the baseline tests will still fail, but I do not want to start working on these unless we're all confident this is the way to go.

teo-tsirpanis commented:
I am still worried by the excessive use of the `box` operator, but I remember that those who care about their code's performance will directly call the `ToString` overload they want.

The core of the issue is that __`FS0670` is overly restrictive and prevents us from doing sensible things__.

My opinion would be to loosen the aforementioned restriction and in the meantime, use the `string` implementation that would be the best _if we assume we are not burdened by `FS0670`_.

As for the changes at hand, they seem better than the previous implementation of the `string` operator (in particular the elimination of the extra `null` check), but the impact of the boxing elimination for structs is not that great, given that some of the most commonly used structs and all custom ones will get eventually boxed.

I am wondering what makes the enums such a special case. I wrote the following C# program

``` csharp
using System;

namespace ToStringTest
{
    static class Program
    {
        static string AnyToString<T>(T x) => x.ToString();
        static void Main()
        {
            Console.WriteLine(AnyToString(521));
            Console.WriteLine(AnyToString(ConsoleKey.U));
        }
    }
}
```

and the results are the expected ones (`521` and `U`). A straightforward call to ToString on a generic type will correctly deduce we are talking about an enum. I understand that the compiler cannot currently distinguish between an integer and an enum backed by an integer, but does a case like `when 'T: Enum = value.ToString()` work?

abelbraaksma commented:
It compiles, but it doesn't work, unfortunately. The reason is that `Enum` is an abstract class, so it will never match.

The problem in F# is that when you use static type checks, any `int` an an enum if that type, are considered equal. The easiest solution would be to drop printing the name of the enum case.

C#, since 7.3, can distinguish between enum and integral types. That's a fairly recent feature. There's not yet a similar proposal for F#, and I don't know how hard it would be to implement (looking at the C# discussions, it wasn't trivial). I assume that's also the reason why your C# example works.

What I haven't tried, though, is if calling `ToString` on either an integral type or enum would have the same result. But then, still, we need to tell the CLR what type it is, and we end up where we started.

I'll think about it some more. 

Btw, it's more than just the null check, that's gone. It's also call vs callvirt, and the second null check (there's no check at all anymore). Casting without boxing is not possible, and calling `IFormattable.ToString` without casting is not possible. That's the conundrum.

Note that in the original, none of these special cases were hit, ever. For instance, using it on a decimal, is now equal to doing `42.42M.ToString(...)` directly.

abelbraaksma commented:
I'll also run some benchmarks. It was already faster, but should be even better now, even for integral types (the ones that are still boxed). The others are as fast as they would be if you'd code it by hand.

abelbraaksma commented:
@teo-tsirpanis, maybe there's a solution after all, even for the last cases that still need boxing now. I was stuck on the idea that we _must_ use `IFormattable`. The reason for this is that we want to pass `InvariantCulture`. However, for integral types (the 8 ints) and for enums, this has no effect. 

If we look at the implementation of `Int32.ToString()` it does indeed use `CurrentCulture`. There are two problems with this: (1) it used to be much slower than InvariantCulture and (2) we cannot ascertain that the output will be equal always for all cultures.

I think that, if the recent rewrites of all `ToString()` functions of the BCL types for .NET Core have had the effect that it is equally fast, we can go with that. Otherwise, the boxing is a small price to pay for those tiny types, compared to the loads of work that is done for converting them to a string.

The idea is to change the 8 integral types to look like this, which compiles, and correctly prints the names of enums:

```f#
when 'T : int32      = (# "" value : 'T #).ToString()
```

I'll do some more testing and create benchmarks for comparison. At least in the past this was much slower than using `InvariantCulture`, let's see how it is now.

abelbraaksma commented:
@teo-tsirpanis, I've given this a little bit more thought. Your C# example works locally by using `ToString()`, but, as mentioned, that won't force the `InvariantCulture` (why on earth that isn't the default on any `ToString()` in the BCL is beyond me, but that's a fact of life).

Consider this:

![image](https://user-images.githubusercontent.com/16015770/86013682-b1ebe200-ba1f-11ea-96cf-ba604ed43e85.png)

Then click <kbd>Apply</kbd> and reset your FSI. This will give:

```f#
> (-12).ToString();;
val it : string = "_12"

> string -12;;
val it : string = "-12"
```

Or with "over line", which apparently is used in some educational systems, so it isn't entirely impossible that this happens in the wild:

```f#
> (-12).ToString();;
val it : string = "‾12"
```

Conclusion: as much as I would like to use `ToString()`, we can't (unless @cartermp/@KevinRansom would jump in and say: _"this is so rare, just use `ToString()` for integral types and be done with it."_).

As far as I know it is not possible to convert to `IFormattable` **and** prevent boxing (I tried, against better knowing, it crashes horribly). The box-less versions for `float` work, however, because it is possible to issue a `call` instruction, since the type is known. For ints, the type is _not known_, I would need to issue a `call` instruction on _either the int, or the concrete enum_. Technically, this information is available during type checking, but not in the stdlib.

teo-tsirpanis commented on src/fsharp/FSharp.Core/prim-types.fs:
```diff
@@ -4465,31 +4465,33 @@ namespace Microsoft.FSharp.Core
              // It is not possible to distinguish statically between Enum and (any type of) int.
              // This way we'll print their symbolic value, as opposed to their integral one 
              // E.g.: 'string ConsoleKey.Backspace' gives "Backspace", rather than "8")
              when 'T : sbyte      = (box value :?> IFormattable).ToString(null, CultureInfo.InvariantCulture)
```
Can't we use the `constrained !!T.` opcode with the `(# ... #)` syntax for the signed integers as well to avoid boxing too? Will it bypass the type system?

abelbraaksma replied:
@teo-tsirpanis, it'll break, because I cannot use `ToString()`, instead I must use `IFormattable::ToString(...)`. For the `IFormattable` call, to bypass boxing, you need to emit the `call` on the exact `type + method`. Since I don't know the type, I cannot choose between, say `System.Int32::ToString(null, InvariantCulture)` and `System.ConsoleKey::ToString(null, InvariantCulture)`. The only alternative is to use `callvirt`. Adding `constrained` doesn't help here, because then it would always resolve to the integer call, instead of enum OR integer.

I've asked several people, but so far nobody has a clue as to how to fix improve this further.

I've updated the original text of this PR with new timings. As you can see there (`UL` type vs `L` type), it certainly would help performance if we could emit such calls.

abelbraaksma commented:
@KevinRansom, I think this is now "as good as it gets". I've updated the original text of the PR and updated the performance charts (as you requested). It's certainly much better now than my first attempt: it's faster, less allocations and better IL code optimizations.

If you think this is good enough to get in, I'll set to work on the baseline tests (or maybe you can help there?).

TIHan commented on src/fsharp/FSharp.Core/prim-types.fs:
```diff
@@ -4457,23 +4459,53 @@ namespace Microsoft.FSharp.Core
+             // It is not possible to distinguish statically between Enum and (any type of) int.
+             // This way we'll print their symbolic value, as opposed to their integral one 
+             // E.g.: 'string ConsoleKey.Backspace' gives "Backspace", rather than "8")
+             when 'T : sbyte      = (box value :?> IFormattable).ToString(null, CultureInfo.InvariantCulture)
```
In order to maintain printing the right thing for enums and not do boxing on int types, I think you can add this case above the int-types:
```fsharp
when 'T : Enum = (retype value : Enum).ToString(null, CultureInfo.InvariantCulture)
```

TIHan replied:
Resolved: @abelbraaksma has tried this but because `Enum` is abstract it won't match on that rule.

abelbraaksma replied:
Indeed ;)

dsyme commented:
@abelbraaksma Why not continute to `inline` the string function, even if SRTP is no longer used?  I don't understand why the `inline` has to be removed.

> that we remove inline from the Operators.string function, but without losing the 'inlinability'. This allows its usage in places where you would now get an error that it "escapes its scope" (report: dotnet/fsharp#7958). Consider:

I believe it is the removal of SRTP that solves this problem for you, rather than the removal of `inline`.

Could you try with the `inline` re-established please?

Thanks

dsyme commented:
@tihan and I spent a while looking at what it would take to optimize

```
   match box value with
   | :? IFormattable as f -> f.ToString(null, CultureInfo.InvariantCulture)
   | _ -> value.ToString()
```
to 
```
 (constrcall<int> arg).ToString(null, CultureInfo.InvariantCulture)
```

We followed the idea of looking at adding information about box/unbox into the optimizer data for expressions:


```
ValueInfo
  | UnknownValue
  | ValValue of ValRef * ExprValueInfo    
  | ConstValue of Const * TType
  ...
  | ValueIsBox of srcTy: TType * tgtTy: TType * ExprValInfo
  | ValueIsUnbox of srcTy: TType * tgtTy: TType * ExprValInfo
```

The code above is roughly this after pattern match compilation and type checking:
```
  let inp  = box arg
  dtree 
    cases:
       IsInst(inp, IFormattable) --> 0
    targets:
       0:   let f = unboxFast<IFormattable>(inp) in f.ToString(null, CultureInfo.InvariantCulture)
```
or perhaps this (pre-binding an isInst call to get 'a is IFormattable')
```
  let inp  = box arg
  let typeTestResult = isInst<IFormattable>(inp)
  dtree 
    cases:
       IsNotNull(typeTestResult) --> 0
    targets:
       0:   let f = typeTestResult in f.ToString(null, CultureInfo.InvariantCulture)
```
The rules we would need to add to the optimizer would be:

1. "Produce ValueIsBox"

       e.g. `let x = box arg  --> x ~ ValueIsBox(int, ValValue(arg, UnknownValue))`

2. "Produce ValueIsUnbox"

       e.g. `unboxFast<ity>(obj) --> ValueIsUnbox(ty, ity, info) when  v ~ ValueIsBox(ty, info) and ty :> ity`

3. "Resolve IsInst using ValueIsBox"

       e.g.  `IsInst(inp, ity)    --> yes when  v ~ ValueIsBox(ty, info) and ty :> ity`

4. "Optimize to unboxed constrcall using ValueIsUnbox"

       e.M(args)           -->  (constrcall<origty> expr-for(info) ).M(args) when e: ty and e ~ ValueIsUnbox(origty, ty, info)

With these we would get  these infos for the example above:

    inp --> ValueIsBox(obj, int, ValValue(arg, UnknownValue))
    f --> ValueIsUnbox(int, IFormattable, ValValue(arg, UnknownValue))

and the overall code would reduce to this:

    (constrcall<int> arg).ToString(null, CultureInfo.InvariantCulture)

However the biggest problem with this is that it involves adding to the ExprValueInfo data structure, which causes compatibility problems for older compilers if the new information is present in nre versions of FSharp.Core. 

One recommendation: We should plan ahead for allowing future additions to ExprValueInfo by changing the failwith in this case of unpickling

```
let rec u_ExprInfo st =
    let rec loop st =
        let tag = u_byte st
        match tag with
        | 0 -> u_tup2 u_const u_ty st |> (fun (c, ty) -> ConstValue(c, ty))
        ...
        | _ -> failwith "loop"  // <--- change this
```
to 
```
let rec u_ExprInfo st =
    let rec loop st =
        let tag = u_byte st
        match tag with
        | 0 -> u_tup2 u_const u_ty st |> (fun (c, ty) -> ConstValue(c, ty))
        ...
        | _ -> UnknownValue
```

We will think on this more :-)

TIHan commented:
Yes, the metadata issue that @dsyme mentioned is present even when trying to add a new constraint to static optimizations.

We will have to think and discuss this more.

abelbraaksma commented:
> I believe it is the removal of SRTP that solves this problem for you, rather than the removal of inline.

The original version had no SRTP, just compiler optimizations, that always led to `anyToString` to be called. Though I didn't test your proposal, if that prevents the error, I'm OK with leaving `inline` in. Though, technically, I don't think it matters, as the generated code will be the same.

abelbraaksma commented:
The proposed optimization path looks promising, and would improve a lot of code. I need to let that sink in a bit. You gave an example with `int`, but I assume you mean that the constrained call will retain the type info, right? Cause otherwise it wouldn't help the underlying info of enum vs integral types.

dsyme commented:
> Though, technically, I don't think it matters, as the generated code will be the same.

I imagine the generated code will be very different at callsites where no inlining happens automatically. But maybe it always happens.

abelbraaksma commented:
@dsyme, the code uses static optimizations. Same approach with `'T` without inline (and not `^T` with inline) is used in other places.

If I follow the compiler flow it appears that these `when` optimizations are always used, and if I'm right, it may help optimize some other Core functions that currently are not inline, or use dynamic tests.

If, for some call, it's not used, either because it doesn't hit the optimize clause, or if the optimizer itself misses the inlining, it uses the generic body of `anyToString`, which is the old behavior (the static `when` clauses are never ever hit in the current version, except the first one, `when struct`).

Just saying, if cases exist where I'm wrong, it works as before, with the same emitted IL, in all other cases, the optimized version will be used.

abelbraaksma commented:
> I believe it is the removal of SRTP that solves this problem for you, rather than the removal of inline.

@dsyme, I can now confirm this is correct. I was under the wrong impression that (1) SRTP means added static constraints, not just presence of `^T` vs `'T` and (2) that `inline` caused FS0670 to be raised.

I've added back the `inline` modifier, but left the tick-T syntax, and indeed, the compiler doesn't balk anymore if the call-site could "escape its scope".

I compared a bunch of different scenarios and the generated IL and couldn't find a difference. There's, however, still the very tiny difference between the SRTP version:

```f#
let anyString x = string x   // before: x gets normalized to obj (unless used), after, it is 'T
```

abelbraaksma commented:
I've also updated the RFC, as I think putting `inline` back in makes it less controversial and can get the proposal more easily accepted: https://github.com/fsharp/fslang-design/pull/479.

abelbraaksma commented:
Merged with #9622, which helps solve the test errors.

abelbraaksma commented:
I didn't really want to make tests in `ExprTests.fs` more incomprehensible than they were, but there you go, did it anyway ;). The reason for adding a differentiation in FSharp.Core versions in the tests is that some jobs run optimization tests **against FSharp.Core 4.5** and against **FSharp.Core 4.6**. This is probably deliberated and prevents backward compat issues.

Found this because I'm now dumping referenced assembly info (see https://dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_apis/build/builds/716771/logs/101):

> Referenced assembly FSharp.Core, Version=4.6.0.0

and:

> Referenced assembly FSharp.Core, Version=4.5.0.0

This also establishes that tests are run on .NET Framework 4.6.1 (again, probably for backward compat reasons) and .NET Core 3.1:

> Test run for D:\a\1\s\artifacts\bin\fcs\Release\net461\FSharp.Compiler.Service.Tests.dll(.NETFramework,Version=v4.6.1)

and 

> Test run for D:\a\1\s\artifacts\bin\fcs\Release\netcoreapp3.1\FSharp.Compiler.Service.Tests.dll(.NETCoreApp,Version=v3.1)

This _may also help to explain the issues @cartermp was encountering when tackling the `not` function, for the bsl tests_. It doesn't really help that the CI reports 16 failing tests when in reality it is 32: one for each old FSharp.Core version. Resulting in: fix one, other one fails, fix other, first one fails, with no glaring difference from looking at the reports.

abelbraaksma commented:
Yessss! Green! This is now ready for review @cartermp.

@dsyme, @tihan, about the remaining optimizations that currently aren't possible, your suggestion is to tackle that through improving optimizations for the general case. 

I'd suggest to merge this and not wait for that (unless you can think of a quick way?), as the `string` function is now as good as it can get (no boxing, optimized IL). The remaining work for the four signed integral types vs enums can go in a new issue when we improve optimizations, or otherwise if `when 'T: enum` is going to be allowed.

cartermp commented:
Since this involved some surprisingly complicated details that @TIHan and @dsyme talked about extensively I'll defer to them on this PR.

abelbraaksma commented:
Hmm, sometimes when I merge I just get a "Merge 'branchX' into branchY", sometimes it copies each commit (like just now, when I merged from the branch of https://github.com/dotnet/fsharp/pull/9622). This muddles the commit history.

Probably not a big deal, but I like it when it just says "Merge X into Y", because that's all the info you usually need.

abelbraaksma commented:
Ping @TIHan, @dsyme. can either or both please review?

As mentioned above, I think this is now "as good as it can get" with current features of the language, and improves performance of `string` by a large margin, plus emits the proper/expected IL.

For the signed-int case, which needs a distinction between `enum` and `intX`, this requires a new feature for the static `when` optimization syntax. I propose to merge this as-is and take the discussion on such improvements into a new thread, as they have wider applicability.

abelbraaksma commented:
> Ping @TIHan, @dsyme. can either or both please review?

/cc @TIHan, @dsyme, :pray:? ;)

cartermp commented:
@abelbraaksma currently we're in a very hard push to finish off and review open type declarations and string interpolation. We're trying to get those completed by the end of this week or maybe the end of next week. After that we should be more "free" :)

abelbraaksma commented:
@cartermp Cool, thanks for the update! If I can help in any way with those tasks =tests or otherwise), ping me :)

cartermp commented:
If you're up for it, our biggest concern right now is ensuring that tooling works well with string interpolation. If you're able to pull down that branch, build a new VS, and use F# scripts to play with the feature and give feedback then that would be great.

Unfortunately it's tricky to use the feature with a real codebase because it relies on preview FSharp.Core functionality that is only accessible in F# scripts or some trickery where a project has its FSharp.Core reference swapped out for the built FSharp.Core. But if you use F# scripts it's fine, which I've been doing.

abelbraaksma commented:
@cartermp Yeah, I know how to cheat a different Core reference (used it in benchmarking for before/after in one run), and how to switch the fsc.exe, I'll give it a shot.
